site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,260 @@
+  end
+
+  def build_jwt(valid_for_minutes = 5)
+    exp = Time.now.to_i + (valid_for_minutes*60)
+    payload = { "iss": "fusionauth.io",
+                "exp": exp,
+                "aud": "238d4793-70de-4183-9707-48ed8ecd19d9",
+                "sub": "19016b73-3ffa-4b26-80d8-aa9287738677",
+                "name": "Dan Moore",
+                "roles": ["USER"]
+    }
+
+    JWT.encode payload, Rails.configuration.x.oauth.jwt_secret, 'HS256'
+
+  end
+end
+```
+
+What we're doing here is writing a test that specifies the JWT will be in the `Authorization` header. Let's look more closely at the token. 
+
+[JWTs have claims](https://tools.ietf.org/html/rfc7519#section-4), basically information embedded in the JWT. The keys of the JSON payload we build in the `build_jwt` function, such as `iss` and `name`, are claims. Some of these are defined in the JWT RFC. These are 'registered' claims. Others are recorded with the IANA but are not part of the standard; these are 'public' claims. And yet others are added to the payload by the token creator; these are 'private' claims. 
+
+
+```ruby
+# ...
+  def build_jwt(valid_for_minutes = 5)
+    exp = Time.now.to_i + (valid_for_minutes*60)
+    payload = { "iss": "fusionauth.io",
+                "exp": exp,
+                "aud": "238d4793-70de-4183-9707-48ed8ecd19d9",
+                "sub": "19016b73-3ffa-4b26-80d8-aa9287738677",
+                "name": "Dan Moore",
+                "roles": ["USER"]
+    }
+
+    JWT.encode payload, Rails.configuration.x.oauth.jwt_secret, 'HS256'
+# ...
+```
+
+Above, we create registered claims that our API may examine. `exp` indicates when the JWT will expire. `aud` is an identifier of who or what this token is intended for (the "audience"). `sub` is the person or piece of software to which this token applies. To quote the RFC: "The claims in a JWT are normally statements about the subject." `iss` is an identifier for the issuer of the JWT. 
+
+We also add the `name` public claim to convey the user's name. `roles` are a private claim with a meaning undefined outside of our application. Note that because the content of JWTs is not typically encrypted, claims should contain no secrets or private data.
+
+The last thing we do is encode our JWT. This signs it, adds needed metadata and creates the URL encoded version. Here's what one of the JWTs generated by `build_jwt` looks like:
+
+```
+eyJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJmdXNpb25hdXRoLmlvIiwiZXhwIjoxNTkwMTgxNjE5LCJhdWQiOiIyMzhkNDc5My03MGRlLTQxODMtOTcwNy00OGVkOGVjZDE5ZDkiLCJzdWIiOiIxOTAxNmI3My0zZmZhLTRiMjYtODBkOC1hYTkyODc3Mzg2NzciLCJuYW1lIjoiRGFuIE1vb3JlIiwicm9sZXMiOlsiVVNFUiJdfQ.P7KXBV8fNElGGr1McKIMQbU7-mZPMxv8tw5AbufZgr0
+```
+
+We use the HMAC signature algorithm because in this tutorial we control both the issuer of the token and the consumer, our API. We can therefore share a secret reliably between them. If we didn't have a good way to share secrets, using an asymmetric key would be a wiser choice.
```

Comment:
Add commas before and after , therefore,
```suggestion
We use the HMAC signature algorithm because in this tutorial we control both the issuer of the token and the consumer, our API. We can, therefore, share a secret reliably between them. If we didn't have a good way to share secrets, using an asymmetric key would be a wiser choice.
```